sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/DefaultSdkMeterProvider.java

```diff
@@ -0,0 +1,122 @@
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * {@code SdkMeterProvider} implementation for {@link MeterProvider}.
+ *
+ * <p>This class is not intended to be used in application code and it is used only by {@link
+ * OpenTelemetry}.
+ *
+ * <p>WARNING: A MetricProducer is stateful. It will only return changes since the last time it was
+ * accessed. This means that if more than one {@link
+ * io.opentelemetry.sdk.metrics.export.MetricExporter} has a handle to this MetricProducer, the two
+ * exporters will not receive copies of the same metric data to export.
+ */
+final class DefaultSdkMeterProvider implements SdkMeterProvider {
+
+  private static final Logger LOGGER = Logger.getLogger(DefaultSdkMeterProvider.class.getName());
+  static final String DEFAULT_METER_NAME = "unknown";
+  private final ComponentRegistry<SdkMeter> registry;
+  private final MeterProviderSharedState sharedState;
+
+  // These are *only* mutated in our constructor, and safe to use concurrently after construction.
+  private final Set<CollectionHandle> collectors = CollectionHandle.mutableSet();
+  private final List<MetricReader> readers = new ArrayList<>();
+
+  DefaultSdkMeterProvider(
+      List<MetricReader.Factory> readerFactories,
+      Clock clock,
+      Resource resource,
+      ViewRegistry viewRegistry,
+      ExemplarSampler exemplarSampler) {
+    this.sharedState =
+        MeterProviderSharedState.create(clock, resource, viewRegistry, exemplarSampler);
+    this.registry =
+        new ComponentRegistry<>(
+            instrumentationLibraryInfo -> new SdkMeter(sharedState, instrumentationLibraryInfo));
+
+    // Here we construct our own unique handle ids for this SDK.
+    // These are guaranteed to be unique per-reader for this SDK, and only this SDK.
+    Supplier<CollectionHandle> handleSupplier = CollectionHandle.createSupplier();
+    for (MetricReader.Factory readerFactory : readerFactories) {
+      CollectionHandle handle = handleSupplier.get();
+      // TODO: handle failure in creation or just crash?
+      MetricReader reader = readerFactory.apply(new LeasedMetricProducer(handle));
+      collectors.add(handle);
+      readers.add(reader);
+    }
+  }
+
+  @Override
+  public MeterBuilder meterBuilder(@Nullable String instrumentationName) {
+    if (instrumentationName == null || instrumentationName.isEmpty()) {
+      LOGGER.fine("Meter requested without instrumentation name.");
+      instrumentationName = DEFAULT_METER_NAME;
+    }
+    return new SdkMeterBuilder(registry, instrumentationName);
+  }
+
+  @Override
+  public CompletableResultCode flush() {
+    List<CompletableResultCode> results = new ArrayList<>();
+    for (MetricReader reader : readers) {
+      results.add(reader.flush());
+    }
+    return CompletableResultCode.ofAll(results);
+  }
+
+  @Override
+  public CompletableResultCode shutdown() {
+    // TODO - prevent multiple calls.
```

Comment:
or just ignore subsequent ones, returning the appropriate result code.

Reply:
Right, do we have an example of that impl I can copy-paste?

Reply:
heh. I looked through the code and found a variety of approaches, none of which look great. It will probably involve saving off the CompletableResultCode somewhere, and just returning it if it's already there.